### ISSUE TITLE
Launch speed is frustratingly slow due to `/etc/vulkan` violations

### ISSUE THREAD

**Issue by JGCarroll** [canonical/steam-snap#390]
### Ensure there isn't an existing issue for this and check the wiki

- [X] This issue is not a duplicate and I have checked the wiki.

### Current Behavior

Running on Ubuntu 24.04 with a RTX 2060 and NVidia drivers 535 with 32bit libraries.

Opening Steam (with no processes running, clean) is frustratingly slow, taking upwards of 1-5 minutes. The apt version launches much quicker.

The snap version  shows error messages relating to being unable to initialize Vulkan, followed by a crashdump generation. The UI eventually appears, after significant delay.

AppArmor shows the following linked denials:
```
Log: apparmor="DENIED" operation="open" class="file" profile="snap.steam.steam" name="/etc/vulkan/explicit_layer.d/" pid=196147 comm="pressure-vessel" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
Log: apparmor="DENIED" operation="open" class="file" profile="snap.steam.steam" name="/etc/vulkan/implicit_layer.d/" pid=196147 comm="pressure-vessel" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
```

### Expected Behavior

The Steam Snap should have access to these directories so that the user experience isn't impacted.

Although I've not tested, I'd also expect this likely breaks games trying to use Vulkan directly (including possibly Proton?).

I've added `/etc/vulkan r,` to the AppArmor profile and reloaded the snap with it. The Vulkan related errors in the console went away, the crashdump doesn't occur, and the login UI appears significantly quicker, on par with the Apt version.

### Steps To Reproduce

1) Open Steam

### Environment

```yaml
os_release:
    name:               "Ubuntu"
    version:            "24.04 LTS (Noble Numbat)"
snap_info:
    steam_revision:     191
    snapd_revision:     21759
lspci:
    09:00.0:            NVIDIA Corporation TU106 [GeForce RTX 2060 Rev. A] (rev a1)
glxinfo:
    gpu:                NVIDIA GeForce RTX 2060/PCIe/SSE2
    gpu_version:        4.6.0 NVIDIA 535.171.04
lscpu:
    model_name:         AMD Ryzen 7 3700X 8-Core Processor
xdg_current_desktop:    ubuntu:GNOME
desktop_session:        ubuntu
```

### gaming-graphics-core22 version

kisak-fresh (default)

### Anything else?

Ultimately this seems like an easy fix of adding `/etc/vulkan r,` to the `steam-support` profile, but I thought I'd raise it here first as there might be nuance I don't understand.

**Comment by thatLeaflet** [comment]
I also have slow startups.

The flatpak also used to start slow, but the [May 7th Client update](https://store.steampowered.com/news/app/593110/view/4182231197207095612) mentioned fixing that. Perhaps the issue also affects the Steam snap but wasn't addressed?

**Comment by ashuntu** [comment]
This is a known issue. We are reworking the app armor profiles (in [snapd](https://github.com/snapcore/snapd)) for the Steam snap pretty heavily this cycle, so at the moment this will remain unresolved until that work is complete. The snap should perform much better after that work is done.

**Comment by zyga** [comment]
This has been fixed in snapd, and will be released in snapd 2.64.

You can give it a try with:
```
sudo snap refresh snapd --edge
```

To go back to a stable release of snapd please run:
```
sudo snap refresh snapd --stable
```

Some changes may need a reboot to apply (not really but easier than precise description).

**Comment by luisalvarado** [comment]
> This has been fixed in snapd, and will be released in snapd 2.64.
> 
> You can give it a try with:
> 
> ```
> sudo snap refresh snapd --edge
> ```
> 
> To go back to a stable release of snapd please run:
> 
> ```
> sudo snap refresh snapd --stable
> ```
> 
> Some changes may need a reboot to apply (not really but easier than precise description).

Thank you. Tested on the following hardware:

```
os_release:
    name:               "Ubuntu"
    version:            "24.04 LTS (Noble Numbat)"
snap_info:
    steam_revision:     200
    snapd_revision:     22641
lspci:
    0000:01:00.0:       NVIDIA Corporation Device 2684 (rev a1)
glxinfo:
    gpu:                NVIDIA GeForce RTX 4090/PCIe/SSE2
    gpu_version:        4.6.0 NVIDIA 555.58.02
lscpu:
    model_name:         13th Gen Intel(R) Core(TM) i9-13900K
xdg_current_desktop:    ubuntu:GNOME
desktop_session:        ubuntu
```

Speed did go up probably 2x or 3x faster for loading compared to the stable snap branch, but it still took a long time to load. To put this into perspective, using a stop watch (Testing 5 times each one and then the average) I got:

Steam Snap - 17 Seconds from executing it to Time to User Interaction
Steam Deb -  3 Seconds from executing it to Time to User Interaction


This was after updating to edge and then rebooting the computer.

**Comment by thatLeaflet** [comment]
Steam snap took 8 seconds from click to first window appearing (warm start, had open previously). Much, much better than before. Felt close to Steam flatpak.

Tested on Ubuntu 24.04.1 release candidate ISO, snapd beta 2.65.1 rev 22830, Steam snap stable 200.

**Comment by JGCarroll** [comment]
I'll close this as there's clearly been huge performance gains for users on the pre-releases, and the relation specifically to `/etc/vulkan` is dealt with so future performance enhancements would be better with their own ticket to track.

For people in this thread, the 2.64 release was effectively skipped for the 2.65 release which contains even further improvements and should be stable in the next week or so.

**Comment by luisalvarado** [comment]
Using snapd 2.65 + Ubuntu 24.10 for anyone reading my last post and this current one (Note that on the last test I was using the Steam version that did not have the family update yet, the current one here has the family update):

Cold Start (5 times) (Cold Start is me rebooting the computer for each test, meaning I had to reboot 15 times, 5 for each case)
Steam Snap - 3 Seconds from executing it to Time to User Interaction
Steam Deb - 3 Seconds from executing it to Time to User Interaction
Steam Flatpak - 6 Seconds from executing it to Time to User Interaction

Warm Start (5 times) (This are running steam again without rebooting. Waiting 30 seconds between running sessions.)
Steam Snap - 3 Seconds from executing it to Time to User Interaction
Steam Deb - 3 Seconds from executing it to Time to User Interaction
Steam Flatpak - 5 Seconds from executing it to Time to User Interaction


Now for the Starting and Closing of Steam (Because yes, closing steam is no picnic) I recorded the times here. The moment it opened visually I closed it with ALT+F4 to trigger the closing process. Here is the results


STEAM DEB (Starting + Closing)

![image](https://github.com/user-attachments/assets/c66f1bb8-1cf5-4022-8960-64bded6c634f)


STEAM SNAP (Starting + Closing)

![image](https://github.com/user-attachments/assets/984f8ae6-6574-4364-8c30-7170c19dfa53)


STEAM FLATPAK (Starting + Closing)

![image](https://github.com/user-attachments/assets/06e82c47-607b-4ea8-ba81-3c6a93ce68d6)
